lib/checks/hasUnsupportedNodes.js

```diff
@@ -1,30 +1,26 @@
-const Node = require('../node');
+const {query} = require('math-nodes');
+
 const resolvesToConstant = require('./resolvesToConstant');
 
 function hasUnsupportedNodes(node) {
-  if (Node.Type.isParenthesis(node)) {
-    return hasUnsupportedNodes(node.content);
-  }
-  else if (Node.Type.isUnaryMinus(node)) {
-    return hasUnsupportedNodes(node.args[0]);
-  }
-  else if (Node.Type.isOperator(node)) {
-    return node.args.some(hasUnsupportedNodes);
-  }
-  else if (Node.Type.isSymbol(node) || Node.Type.isConstant(node)) {
-    return false;
-  }
-  else if (Node.Type.isFunction(node, 'abs')) {
-    if (node.args.length !== 1) {
-      return true;
+  if (query.isApply(node)) {
+    if (query.isAbs(node)) {
+      if (node.args.length !== 1) {
+        return true;
+      }
+      if (!resolvesToConstant(node.args[0])) {
+        return true;
+      }
     }
-    if (node.args.some(hasUnsupportedNodes)) {
-      return true;
+    else if (query.isNthRoot(node)) {
```

Comment:
It's probably worth having some sort of validation function to check that ASTs constructed manually are valid.